src/trace-plugin-loader.ts

```diff
@@ -16,278 +16,485 @@
+              }
             }
           }
-          const rootPatch = patchSet[''];
-          if (rootPatch && isInternalIntercept(rootPatch)) {
-            return rootPatch.module;
-          } else {
-            return null;
-          }
         }
+        return exportedValue;
+      });
+      this.internalState = PluginLoaderState.ACTIVATED;
+      this.logger.info(`PluginLoader#activate: Activated.`);
+    } else if (this.internalState === PluginLoaderState.DEACTIVATED) {
+      throw new Error('Currently cannot re-activate plugin loader.');
+    } else {
+      throw new Error('Plugin loader already activated.');
+    }
+    return this;
+  }
 
-        function moduleAlreadyPatched(
-            instrumentation: InternalPlugin, moduleRoot: string) {
-          return instrumentation.patches[moduleRoot];
-        }
+  /**
+   * Deactivates the plugin loader, preventing additional plugins from getting
+   * loaded or applied, as well as unpatching any modules for which plugins
+   * specified an unpatching method.
+   */
+  deactivate(): PluginLoader {
+    if (this.internalState === PluginLoaderState.ACTIVATED) {
+      // Unpatch the unpatchable functions.
+      for (const pluginWrapper of this.pluginMap.values()) {
+        pluginWrapper.unapplyAll();
+      }
+      this.internalState = PluginLoaderState.DEACTIVATED;
+      this.logger.info(`PluginLoader#deactivate: Deactivated.`);
+    }
+    return this;
+  }
 
-        // Future requires get patched as they get loaded.
-        // tslint:disable:no-any
-        return function Module_load(
-            this: any, request: string, parent?: NodeModule,
-            isMain?: boolean): any {
-          // tslint:enable:no-any
-          const instrumentation = plugins[request];
-          if (instrumentation) {
-            const moduleRoot = util.findModulePath(request, parent);
-            const moduleVersion =
-                util.findModuleVersion(moduleRoot, originalModuleLoad);
-            if (moduleAlreadyPatched(instrumentation, moduleRoot)) {
-              return originalModuleLoad.apply(this, arguments);
-            }
-            logger.info('Patching ' + request + ' at version ' + moduleVersion);
-            const patchedRoot =
-                loadAndPatch(instrumentation, moduleRoot, moduleVersion);
-            if (patchedRoot !== null) {
-              return patchedRoot;
-            }
-          } else {
-            const modulePath =
-                Module._resolveFilename(request, parent).replace('/', path.sep);
-            if (intercepts[modulePath]) {
-              return intercepts[modulePath].interceptedValue;
-            }
-          }
-          return originalModuleLoad.apply(this, arguments);
-        };
-      });
-}
+  /**
+   * Adds a search path for plugin modules. Intended for testing purposes only.
+   * @param searchPath The path to add.
+   */
+  static setPluginSearchPath(searchPath: string) {
```

Comment:
I'd use a rather long and ugly name for a testing-only method. Something like `setPluginSearchPathForTestingOnly()`.

Reply:
Done